### ISSUE TITLE
[🔌 Provider]: SNMP provider

### ISSUE THREAD

**Issue by shahargl** [keephq/keep#2112]
Send SNMP traps/events into Keep as alerts

**Comment by Matvey-Kuk** [comment]
Labeling this one as a Good First Issue.

Use https://docs.keephq.dev/providers/adding-a-new-provider as a starting point :)

If you want to work on this, join Slack and we will help: https://slack.keephq.dev/ !

**Comment by michael-azogu** [comment]
/attempt #2112

im looking into this

**Comment by shahargl** [comment]
Hey @onyedikachi-david, do you have any idea how to implement it? Any experience with SNMP?

**Comment by michael-azogu** [comment]
@shahargl from my research...
its basically like Prometheus for network devices.
traps (like os interrupts) are pushed when certain events occur with a context payload.
I need to figure out how to parse & whether ill need pysnmp client

**Comment by onyedikachi-david** [comment]
@shahargl Not in SNMP in particular, but have a good experience with opentelemetry/observability, I am currently extending an instrumentation by adding event-based tracking implementation across providers for a Open-source observability platform for LLM applications (openllmetry).

**Comment by shahargl** [comment]
Ok let’s try! Assigned

**Comment by Matvey-Kuk** [comment]
This issue is up for grabs :)

**Comment by c0d33ngr** [comment]
/attempt #2112

Let me see what I can do here

**Comment by Omar8345** [comment]
@shahargl Does Algora have PayPal payouts? If not, is there manual PayPal payout from keep directly?

**Comment by rahulrajesh21** [comment]
@shahargl  Can you review my pr.

**Comment by advenk** [comment]
/attempt #2112

i'd like to give this a shot if its still open.

**Comment by Harshvardhan-91** [comment]
@shahargl can i work on this?

**Comment by Kesavaraja67** [comment]
Hi @shahargl, seeing that the previous attempts from September seem inactive, I would like to pick this up.

Plan: I will implement the SnmpProvider using pysnmp to handle trap reception and parsing, converting them into Keep alerts.

/attempt #2112

**Comment by matiasmagni** [comment]
/attempt #2112

### Implementation Plan:

1. Analyze existing optimization functions and models in the Python backend to understand current metric weighting and identify potential extension points.
2. Review existing data models (e.g., Plan or configuration) to determine where new optimization preferences (quality, speed, resource) could be stored.
3. Modify or create new Python backend functions to accept an optimization_target parameter.
4. Implement logic in the backend to dynamically adjust `metric_weights` based on the chosen optimization_target.
5. If "resource" optimization is a new concept, define and implement a mechanism to measure and incorporate a "resource" metric into the optimization scoring.
6. Update backend API endpoints to receive the new optimization_target parameter from the frontend.
7. Add UI components (e.g., radio buttons, dropdown) in the relevant React/TypeScript frontend section for users to select "Optimize for Quality," "Optimize for Speed," or "Optimize for Resource."
8. Integrate the new UI controls with the frontend's state management and ensure the selected optimization target is sent to the backend during API calls.
9. If necessary, update TypeScript interfaces (IPlan, IPlanStep, trainingParams) to accommodate the new optimization target parameter.
10. Write unit tests for the updated backend optimization logic.
11. Develop integration tests to verify the end-to-end functionality, from UI selection to backend processing, for each optimization target.
12. Implement UI tests to ensure new frontend components behave as expected.
13. Update relevant documentation (API documentation, user guides) to reflect the new optimization parameters.

**Comment by matiasmagni** [comment]
I have addressed this issue, @shahargl I need push permissions for my branch: `feat/issue-2112-optimize-ai-settings`

**Comment by natkon20043** [comment]
/attempt #2112 
Hi guys I wanna try this too its my first bounty so hopefully I can get adjusted to everything well and this seems like a nice place to start lol

**Comment by rahim890321-boop** [comment]
"I would like to work on this issue. I have experience with UI components and can improve the user identification in the side navigation bar as requested. Please assign this to me."

**Comment by andynewtw** [comment]
/attempt #2112

I'm claiming this bounty! I'll implement an SNMP Trap receiver that converts SNMP traps into Keep alerts.

**Implementation Plan:**
1. Create `snmp_provider/snmp_provider.py` with webhook endpoint for SNMP trap forwarding
2. Use `pysnmp` or similar library for SNMP trap parsing
3. Map SNMP trap data to AlertDto fields
4. Support both SNMPv2c and SNMPv3 traps
5. Add documentation for how to configure SNMP trap forwarding to Keep

**Technical Approach:**
- Since SNMP traps are UDP-based, I'll create a webhook endpoint that accepts forwarded traps from an SNMP trap receiver (like snmptrapd)
- The provider will parse trap data and convert to Keep's AlertDto format
- Will include severity mapping based on SNMP trap OID

ETA: 2-3 hours

---
*This bounty is being worked on by an AI agent (死龍蝦/Dead Lobster) using GitHub Copilot CLI.*

**Comment by Harmatta** [comment]
Proposed implementation (SNMP provider):\n- Webhook-based ingestion endpoint /alerts/event/snmp for trap JSON payloads (v1/v2c/v3).\n- Provider class keep/providers/snmp_provider/snmp_provider.py (BaseProvider).\n- Parse version/enterprise/agent/trapType/varbinds; map OIDs -> fields; severity mapping; linkUp/linkDown resolution.\n- Configurable OID->severity mapping via ProviderConfig.\n- Tests for v1/v2c/v3 parsing + resolution + severity.\nIf this approach is acceptable, I’ll submit a PR.

**Comment by HuzyafaK** [comment]
/attempt #2112

I am implementing a Native Async SNMP Trap Listener that aligns with Keep's mission of reducing alert noise and providing a "single pane of glass" for infrastructure monitoring.

The Value-Add:

Standalone Operation: Unlike previous attempts that rely on external HTTP forwarders, my implementation is a native Python UDP listener. This makes Keep a true standalone SNMP manager.

AIOps Signal Denoising: I am building in "Flapping Detection." Using a time-window sliding buffer, the provider will automatically deduplicate rapid-fire traps (e.g., oscillating interfaces), preventing alert fatigue before it reaches the UI.

Smart Severity Mapping: Direct mapping of well-known Traps to Keep's internal AlertDto severities (e.g., linkDown -> CRITICAL).

This makes the SNMP provider "Enterprise-Ready" out of the box with zero external configuration.

**Comment by bamontejano** [comment]
Hi maintainers. Just a friendly reminder that a complete and working implementation for this SNMP provider bounty is already submitted and awaiting review in PR #5599. It includes a native listener and is ready for production. We are happy to address any feedback to get it merged. Thank you.

**Comment by saimuren** [comment]
I'd like to implement the SNMP provider. I have experience with network monitoring and SNMP (v2c/v3).

### Implementation Plan
1. **SNMP Polling**: Use `pysnmp` for SNMP GET/WALK operations
2. **Trap Receiver**: Listen for SNMP traps and convert to Keep alerts
3. **MIB Support**: Parse standard MIBs for human-readable OID translation
4. **Configuration**:
   - Community string (v2c) / USM credentials (v3)
   - Target hosts and OIDs to monitor
   - Polling interval
   - Trap listener port

### Provider Interface
```python
class SnmpProvider(BaseProvider):
    def validate_config(self): ...
    def query(self, oids: list) -> dict: ...
    def setup_trap_listener(self): ...
```

Ready to start immediately.

**Comment by tianzhuxhc-bit** [comment]
Hi maintainers, I'd like to work on this bounty.

Plan:
1) Reproduce and align scope quickly
2) Submit an initial PR with tests/docs
3) Iterate fast based on review feedback

I can post first progress within 12 hours.

**Comment by franamaro-dev** [comment]
/attempt #2112

Implementation Plan:
I plan to implement an SNMP trap listener as a new Inbound Provider.

Create a new SnmpInboundProvider class using pysnmp-lextudio.

Implement a background task/server to listen on UDP port 162.

Map SNMP Variable Bindings (VarBinds) to Keep's alert schema.

Ensure it runs within the Docker environment and can be toggled via configuration.

**Comment by pratik-programmer** [comment]
Hi! I'm interested in working on this issue. 
I'd like to implement the SNMP provider for Keep. 
Can you assign this to me?

**Comment by natkon20043** [comment]
/claim #2112 , my pr is #5509

**Comment by zhaog100** [comment]
Hi! I'd like to work on this bounty. I'll submit a PR.

**Comment by krishnakumarsethi** [comment]
/attempt #5495 

🏗️ 1. Project Setup
Locate Directory: All provider code resides in keephq/keep/providers/. Create a new directory: snmp_provider/.

Dependencies: Add pysnmp-lextudio (the modern, maintained fork of pysnmp) to the project's requirements to handle the SNMP protocol logic.

⚙️ 2. Define the Provider Class
Create snmp_provider.py and inherit from the base BaseProvider class.

Authentication: Define the required configuration fields (e.g., host, port, community_string for v2c, or user/authKey/privKey for v3).

Connection: Implement the __init__ method to initialize the SNMP engine.

🔍 3. Implement Polling (Query)
Implement the query method to allow Keep workflows to fetch data from devices.

Logic: Use SNMP GET or WALK requests to retrieve specific OIDs (Object Identifiers).

Standardization: Map the raw SNMP response into a structured dictionary that Keep can process as an alert or event.

📥 4. Implement Trap Listener (Ingestion)
Since SNMP traps are "pushed" from devices, this requires a background listener.

Listener: Set up an async UDP server (typically on port 162 or 163) using pysnmp's NotificationReceiver.

Webhook Integration: Ideally, the listener should forward received traps to Keep's internal alert ingestion endpoint (/alerts/event) so they appear in the UI automatically.

OID Mapping: Include a basic mapping logic to convert numeric OIDs (e.g., .1.3.6.1...) into human-readable strings if possible.

🧪 5. Verification & Demo
Unit Tests: Mock SNMP responses to ensure the provider handles different OID types and authentication errors.

Demo Video: As required by the [bounty rules](https://github.com/keephq/keep/issues/2112), record a short clip showing:

Connecting the SNMP provider in the Keep UI.

Successfully "walking" a test device or receiving a simulated SNMP trap.

**Comment by CuboYe** [comment]
/claim 2112

Claiming territory for a scoped fix attempt now. If maintainer preference differs, I can adapt quickly.

**Comment by maketes79** [comment]
Hi team, I’m interested in working on #2112.
/attempt #2112

**Comment by maketes79** [comment]
Implementation is complete and local/manual validation passed:
focused SNMP provider tests passed
single linkDown webhook test returned 202 Accepted
single linkUp webhook test returned 202 Accepted
batch JSON webhook test returned 202 Accepted

Could a maintainer please approve the pending workflows and authorize the required checks for this fork PR if everything looks good?

This PR closes #2112 and includes the bounty claim.

**Comment by dei96303-boop** [comment]
> /bounty 200

Hi, I have implemented the SNMP provider as requested in #2112. I've also signed the CLA. Please let me know if any further changes or tests are required. Thanks!

**Comment by chengyixu** [comment]
/claim #2112 via PR #6168

Implemented a complete SNMP trap provider:
- Webhook-only (no UDP listener), bridges snmptrapd/snmptt → Keep HTTP
- Normalises all SNMPv1 generic trap types (0-6) to severity + status
- Explicit severity field + keyword heuristic fallback
- 55 unit tests, MDX docs, zero external SNMP library dependencies

PR: https://github.com/keephq/keep/pull/6168

**Comment by chengyixu** [comment]
/attempt #2112

Submitted PR #6172 with full SNMP provider:
- Push mode (UDP trap receiver + snmptrapd webhook) + Pull mode (OID polling)
- SNMPv1/v2c/v3 support, 112 unit tests, configurable OID severity mapping

**Comment by shehryarakif** [comment]
/attempt #2112
 Configuration Guide: SNMP Provider in Keep
The SNMP provider supports SNMP v1, v2c, and v3 protocols, with automatic severity mapping for standard SNMP traps. The process involves configuring the provider within Keep.

1. Add the Provider in Keep
Navigate to the Keep dashboard and add a new provider. Search for and select "SNMP."

2. Configure the Provider
Configure the provider based on the SNMP version used by your network devices. Choose between a basic setup (v1/v2c) or an advanced one (v3):

Basic Configuration (SNMP v1/v2c)

listen_address: 0.0.0.0 (listens on all interfaces)

listen_port: 162 (standard SNMP trap port)

community_string: public (match your device's community string)

Advanced Configuration (SNMP v3)

listen_address: 0.0.0.0

listen_port: 162

security_name: snmpuser

auth_protocol: SHA (or MD5)

auth_key: your_auth_password

priv_protocol: AES (or DES)

priv_key: your_privacy_password

3. Configure Your Network Devices
Ensure your network devices (e.g., routers, switches, firewalls) are configured to send SNMP traps to the IP address and port where your Keep instance is running.

💡 How It Works
Once configured, the SNMP provider listens for traps on the designated port. When a trap is received, it automatically maps standard SNMP traps (like linkDown, linkUp, authenticationFailure) to appropriate alert severities and enriches the alert with meaningful context.

A generated alert in Keep will typically include:

Severity: Automatically mapped from the trap data.

Source: The IP address of the network device that sent the trap.

Description: Detailed trap information, including the system name, uptime, and interface details.

Current State of SNMP in Keep
Keep already has an SNMP provider (providers/snmp/). It supports listening for traps and polling via GET. The "Good First Issue" likely refers to one of these improvements:

Enhancing trap MIB parsing (e.g., adding custom MIB support)

Adding SNMP v3 polling support (currently partial)

Improving test coverage for the SNMP listener

Creating a dedicated example/demo for the docs

If a new contributor wants to work on this, here's the roadmap based on your link.

🛠️ Guide for New Contributors: Extending the SNMP Provider
1. Start with the official provider dev guide:
[Adding a New Provider – Keep Docs](https://docs.keephq.dev/providers/adding-a-new-provider)

2. Check the existing SNMP provider code:

Location: keep/providers/snmp_provider/

Entry point: snmp_provider.py (defines SNMPProvider class)

Trap handler: Uses pysnmp to listen for UDP traps on port 162.

3. Identify the specific task:
Look for issues labeled good first issue with provider/snmp tags in the [Keep GitHub repo](https://github.com/keephq/keep).

4. Set up a dev environment:

bash
git clone https://github.com/keephq/keep.git
cd keep
poetry install
# Run a local SNMP trap simulator for testing
sudo poetry run python examples/simulate_snmp_trap.py
5. Join Slack for real-time help:
[keephq.slack.com](https://slack.keephq.dev/) – drop into #contributors and mention you're tackling the SNMP issue.

/claim #2112

**Comment by Ogak-AI** [comment]
/attempt #2112 

I plan to implement the SNMP provider by adding a new module under keep/providers that uses pysnmp to query configured OIDs. I’ll ensure it follows the same interface as existing providers, add configuration options for host, port, and community string, and handle errors gracefully. Tests will include mocked SNMP responses and a simulator for integration. Documentation will cover setup and usage, and I’ll provide a short demo video showing metrics retrieval.

**Comment by ska2704** [comment]
/attempt #2112

Working on this. Implementation plan:
- Push-only provider (no auth needed, devices send traps to Keep's webhook)
- `_format_alert()` converts trap payload to AlertDto
- Supports generic_trap types 0-6 with correct severity mapping
- 6 realistic mock scenarios in alerts_mock.py
- No external dependencies beyond Keep's existing stack

**Comment by muhammadsaadkhatri33-eng** [comment]
I plan to implement the SNMP Trap provider using pysnmp-le. I will create a listener on port 162 that captures incoming traps and maps their OIDs to Keep's AlertDto. I'll include support for SNMP v1/v2c and ensure proper severity mapping for standard traps like linkDown/linkUp.

**Comment by Leviann** [comment]
/attempt #2112

Plan (delivered in PR keephq/keep#6265):
1. Add \snmp\ provider module discovered via \*_provider\ layout.
2. Accept JSON trap payloads on \POST /event/snmp\ (forwarder-friendly); map to \AlertDto\ + standard OID severity hints.
3. Unit tests for parse + format paths.

Branch: \eat/snmp-provider-2112\ on fork Leviann/keep.

**Comment by vasquezmyguy-collab** [comment]
/attempt #2112

PR opened: https://github.com/keephq/keep/pull/6276

Implementation includes:
- SNMP v1/v2c/v3 support
- Consumer pattern (start_consume/stop_consume)
- pysnmp-lextudio v6 asyncio API
- Severity mapping for common traps
- Mock alert data

**Comment by vasquezmyguy-collab** [comment]
/attempt #2112

PR opened: https://github.com/keephq/keep/pull/6280

Implementation includes:
- SNMP v1/v2c/v3 polling via pysnmp hlapi asyncio
- SNMP trap receiver (background daemon thread)
- Configurable auth/priv protocols (MD5, SHA, DES, AES variants)
- Well-known trap OID severity mappings
- Standard Keep provider interface
- Mock alert data for testing

**Comment by junaiddshaukat** [comment]
/attempt #2112
Hi @Matvey-Kuk @shahargl , I'd like to take a shot at this one.
Here's my plan:
- Webhook-based provider, similar in shape to the Prometheus and Grafana providers. User runs `snmptrapd` (standard Linux/macOS daemon) with a tiny exec hook that posts the trap as JSON to Keep's webhook endpoint.
- `SnmpProvider._format_alert()` parses the trap payload (trap OID, varbinds, source host, community) into an `AlertDto`.
- A default OID → severity map for well-known traps (linkDown, coldStart, authenticationFailure, etc.) that users can override.
- Small `snmptrapd.conf` snippet + bridge script under `examples/snmp/` so users can be up and running in a couple of minutes.
- Provider icon, docs page under `docs/providers/documentation/snmp-provider.mdx`, unit tests, and a short demo video.
Going with the webhook shape instead of a native UDP listener because port 162 is privileged and a per-tenant listener doesn't fit Keep's architecture — offloading the raw SNMP handling to `snmptrapd` keeps the provider simple and testable.

**Comment by thinkflow-hub** [comment]
/attempt

ThinkFLOW HUB is ready to harden Keep with a production-grade SNMP Provider. 

We don't just "add features" - we build high-reliability integrations that survive the most demanding infrastructure. 
Our team specializes in autonomous swarms and enterprise automation, and we are bringing that same "zero-trust" engineering to Keep.

The plan:
1. Implement a robust SNMP trap/event listener based on your provider standards.
2. 2. Stress-test against real-world SNMP v2c/v3 scenarios.
3. 3. Deliver a clean, audited PR that is ready for production.
Check our work: https://github.com/thinkflow-hub/OpenClaw-Portfolio

Let's get this done.

**Comment by thinkflow-hub** [comment]
/attempt #6117 
ThinkFLOW HUB is ready to harden Keep with a production-grade SNMP Provider.

We don't just "add features" - we build high-reliability integrations that survive the most demanding infrastructure. Our team specializes in autonomous swarms and enterprise automation, and we are bringing that same "zero-trust" engineering to Keep.

The plan:
1. Implement a robust SNMP trap/event listener based on your provider standards.
2. 2. Stress-test against real-world SNMP v2c/v3 scenarios.
3. 3. Deliver a clean, audited PR that is ready for production.
Check our work: https://github.com/thinkflow-hub/OpenClaw-Portfolio

Let's get this done.

**Comment by thinkflow-hub** [comment]
/attempt #6117 
ThinkFLOW HUB is ready to harden Keep with a production-grade SNMP Provider.

We don't just "add features" - we build high-reliability integrations that survive the most demanding infrastructure. Our team specializes in autonomous swarms and enterprise automation, and we are bringing that same "zero-trust" engineering to Keep.

The plan:
1. Implement a robust SNMP trap/event listener based on your provider standards.
2. 2. Stress-test against real-world SNMP v2c/v3 scenarios.
3. 3. Deliver a clean, audited PR that is ready for production.
Check our work: https://github.com/thinkflow-hub/OpenClaw-Portfolio

Let's get this done.

**Comment by manja316** [comment]
/attempt #2112

Submitted PR #6289 with a complete SNMP provider implementation:
- Webhook-based provider that receives SNMP trap data as JSON
- Automatic severity/status mapping from SNMP generic trap types
- Support for SNMPv1/v2c/v3 formats
- Documentation with snmptrapd and pysnmp setup guides

**Comment by alexanderxfgl-bit** [comment]
I'd like to work on this. I have experience building SNMP integrations and can implement a full SNMP provider with trap reception, MIB parsing, and polling support for Keep's alerting framework.

**Comment by nehemiyawicks** [comment]
/attempt #2112

I've studied the 5 previous PRs (especially #4884 which got closest) and the Keep provider codebase. Here's why those failed and how I'll address each issue:

**Why PR #4884 was rejected (and my fixes):**

1. **Event loop management** - Previous PR used `asyncio.new_event_loop()` + `loop.run_forever()` in a thread, causing conflicts with Keep's async context. I'll use pure threading with `socket` for UDP or pysnmp's synchronous API, avoiding asyncio entirely in the trap listener thread.

2. **E2E tests used `subprocess.run`** - Maintainer explicitly asked for docker-compose. I'll add a `docker-compose.yml` with an SNMP trap sender container and integrate it with the existing e2e test infrastructure using Playwright.

3. **`entrypoint.sh` installing pip packages** - Previous PR modified the entrypoint to install pysnmp at runtime. I'll add `pysnmp-lextudio` to `pyproject.toml` dependencies instead.

4. **Thread shutdown not clean** - `dispose()` couldn't reliably stop `loop.run_forever()`. I'll use a `threading.Event()` for clean shutdown signaling and a non-blocking UDP socket with timeout.

**Implementation plan:**

- Provider class extending `BaseProvider` with `start_consume()` for push-based trap ingestion
- SNMPv1/v2c/v3 auth config via pydantic dataclass (community string, USM credentials)
- Map SNMP trap fields (OID, varbinds, source IP, timestamp) to Keep alert schema
- `validate_scopes()` checking UDP port availability
- Unit tests for trap-to-alert mapping
- Docker-compose based e2e tests per maintainer requirements

**Comment by orbitwebsites-cloud** [comment]
/attempt #2112

**Implementation plan:**
Implement SNMP provider integration using the provided documentation and Slack support, sending SNMP traps/events into Keep as alerts.

Will submit a PR shortly.

**Comment by orbitwebsites-cloud** [comment]
/attempt #2112

**Implementation plan:**
Implement SNMP provider integration using the provided documentation and Slack support.

Will submit a PR shortly.

**Comment by orbitwebsites-cloud** [comment]
/attempt #2112

**Implementation plan:**
Implement SNMP provider integration using the provided documentation and Slack support.

Will submit a PR shortly.

**Comment by lovetovisitLLCdev** [comment]
/attempt #2112

Implementing SNMP provider with trap receiver (push) + device polling (pull) + SNMPv3 USM support. PR submitted at #6390.

**Comment by Naitik120gupta** [comment]
can you assign this to me?

**Comment by aadityaranjan01** [comment]
/attempt #2112

PR submitted at #6391

**Comment by vasquezmyguy-collab** [comment]
/attempt #2112

I'll implement the SNMP provider for Keep. Planning to use pysnmp-lewix (pure Python, no system deps) to handle SNMP trap listening and MIB parsing, following the existing provider pattern. Will open a PR with /claim once implementation is ready.

**Comment by vasquezmyguy-collab** [comment]
/attempt #2112

PR opened: https://github.com/keephq/keep/pull/6399

**Comment by irisdigitaldocs-maker** [comment]
Submitted a webhook-based implementation in #6400 — zero new dependencies, 30+ tests, full MDX docs. Happy to iterate on any feedback!

**Comment by jofreyn7** [comment]
Hi, I'd like to work on this bounty. I'm planning to implement an SNMP v1/v2c trap listener using pysnmp that receives traps and converts them to Keep AlertDto objects. Will submit a PR within 2 days

**Comment by brone1323** [comment]
/attempt #2112

Implementation plan: Native SNMP trap listener (not webhook), modeled after the Kafka consumer-style provider.

- `keep/providers/snmp_provider/snmp_provider.py` extending `BaseProvider`
- `pysnmp` `AsyncoreDispatcher` + `ntfrcv.NotificationReceiver` for v1/v2c trap reception
- Trap OID + var binds ? `AlertDto` mapping with severity heuristic over OID/var values
- v3 explicitly rejected (USM key/auth wiring deferred)
- 11 unit-test cases, MDX + autogenerated docs
- `pysnmp = ^4.4.12` added to `pyproject.toml`

PR: https://github.com/keephq/keep/pull/6430
Differentiator vs PR #6400: real listener vs webhook shim.

**Comment by ppra81** [comment]
PR submitted: https://github.com/keephq/keep/pull/6437
/claim #2112

**Comment by phquand2000** [comment]
__acb_probe__

**Comment by rafael81** [comment]
/attempt #2112

I will take a focused pass on a production-fit SNMP provider rather than another thin duplicate. Plan:

- inspect the latest open SNMP PRs and maintainer feedback first, so this does not repeat rejected approaches
- implement the smallest provider shape that fits Keep's existing provider lifecycle and alert DTO mapping
- prefer a testable trap-to-alert conversion core, with docs and demo video as required by the bounty
- include `/claim #2112` only in the PR body after there is working code and validation

This work will be AI-assisted and manually reviewed.

**Comment by danichese** [comment]
/attempt #2112

I will implement a narrow webhook-first SNMP provider following Keep's existing provider patterns. I will inspect current open SNMP PRs first to avoid duplicating rejected approaches, then add focused tests/docs and submit a PR with validation details.

**Comment by itsbryanman** [comment]
/attempt #2112

Going native UDP listener (consumer provider w/ pysnmp-lextudio asyncio API), not the webhook-forwarder route most of the prior PRs took. One provider covers v1/v2c/v3, OID→severity map with user override, fingerprint pairs linkDown/linkUp so resolution works. Tests + docs in the PR.

**Comment by shahargl** [comment]
N/A
